lsy_drone_racing/envs/real_race_env.py

```
@@ -289,36 +271,130 @@ def send_action(self, action: NDArray):
             # TODO: The estimators can't handle state commands, so we simply don't send anything
             # Make sure to use the legacy estimator with the state interface
 
-    def _connect_to_drone(self, radio_id: int, radio_channel: int, drone_id: int) -> Crazyflie:
+    def _update_track_poses(self, deploy_config: ConfigDict, track: ConfigDict):
+        """Update the track poses from the motion capture system."""
+        if deploy_config.get("real_track_objects", True):
+            tf_names = [f"gate{i}" for i in range(1, self.n_gates + 1)]
+            tf_names += [f"obstacle{i}" for i in range(1, self.n_obstacles + 1)]
+            try:
+                ros_connector = ROSConnector(tf_names=tf_names, timeout=10.0)
+                pos, quat = ros_connector.pos, ros_connector.quat
+            finally:
+                ros_connector.close()
+            try:
+                for i in range(self.n_gates):
+                    self.gates.pos[i, ...] = pos[f"gate{i + 1}"]
+                    self.gates.quat[i, ...] = quat[f"gate{i + 1}"]
+                for i in range(self.n_obstacles):
+                    self.obstacles.pos[i, ...] = pos[f"obstacle{i + 1}"]
+            except KeyError as e:
+                logger.error(
+                    f"Could not find track objects in the ROS TF tree: {e}. \
+                             Have you enabled the track objects in Vicon/ \
+                             started the motion capture tracking node?"
```

```suggestion
                    f"Could not find track objects in the ROS TF tree: {e}. Have you enabled the "
                    "track objects in Vicon and started the motion capture tracking node?"
```

---

It's also not clear why you would want to catch, log, and then reraise the error. Were students unsure what was happening?
In general, try to use as few try-except blocks as possible. If something fails early that's a good thing. There are exceptions of course, e.g. the try finally around the connector to prevent processes hanging indefinitely